library/core/src/ptr/const_ptr.rs

```diff
@@ -463,6 +464,21 @@ impl<T: ?Sized> *const T {
     #[stable(feature = "pointer_byte_offsets", since = "1.75.0")]
     #[rustc_const_stable(feature = "const_pointer_byte_offsets", since = "1.75.0")]
     #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
+    #[requires(
+        // If count is zero, any pointer is valid including null pointer.
+        (count == 0) ||
+        // Else if count is not zero, then ensure that adding `count` doesn't cause 
+        // overflow and that both pointers `self` and the result are in the same 
+        // allocation 
+        (mem::size_of_val_raw(self) != 0 &&
```

Comment:
One last thing: I don't think you need the `mem::size_of_val_raw(self) != 0` check on these anymore.

Reply:
But count should not be greater than 0 for ZSTs, right?

Reply:
Not necessarily. Just because the referent is a ZST, doesn't mean the allocation it's derived from is zero-sized; e.g.:
```rust
fn main() {
    let allocation = &[0u8; 64];
    let first = allocation as *const _ as *const ();
    // SAFETY: `first` points to a ZST at the first address
    // in `allocation`. `second` points to the second address
    // in that same allocation. 
    let second = unsafe { first.byte_offset(1) };
}
```

Reply:
@jswrenn That's right, I get your point. Then how can we ensure that if a pointer is pointing to ZST, then the count should always be 0?

Reply:
@jswrenn , updated the contracts to handle this case. I also added two new harnesses that test:
* if the contracts panic upon having count != 0 when the allocation is zero-sized.
* if the contracts work when the pointer is ZST but is derived from a non-zero sized allocation.